### PR TITLE
Add the ability to do a multi-configuration build to the kernel unit tests

### PR DESCRIPTION
**Pull request by alfred2g**
Add the ability to do a multi-configuration build to the kernel unit tests

Description
-----------
Add a model Makefile that is able to build multi tests with multi-configurations
add tests of the form tasks_utest_1.c with FreeRTOSConfig_1.h and tasks_utest_2.c with FreeRTOSConfig_2.h

Test Steps
-----------
make tasks builds with no errors or warnings

Note: must add tasks to the UNITS variable for this to work
```makefile
UNITS := list tasks
```
then run
```bash
$ make tasks
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: FreeRTOS/Test/CMock/tasks/Makefile [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+     $(shell mkdir -p  $(MOCK_DIRS))
+     $(shell for n in $(MOCK_DIRS) ; do touch -a $(n)/mock_dummy.c)
+endif
+
+EXECUTABLES := $(addprefix $(BIN_DIR)/,$(patsubst %.c,%,$(SOURCES)))
+EXEC_PATH := $(addprefix $(BIN_DIR)/,$(EXEC))
+
+.PHONY: all Makefile directories
+
+.SECONDEXPANSION:
```
Comment (paulbartell) [inline FreeRTOS/Test/CMock/tasks/Makefile:77]: I'm not familiar with how SECONDEXPANSION / SECONDARY works. It seems to build correctly though (once I replace mkdir with mkdir -p).

**Review by paulbartell — CHANGES_REQUESTED** [review]

File: FreeRTOS/Test/CMock/tasks/Makefile (hunk 1 above)
Comment (alfred2g) [reply FreeRTOS/Test/CMock/tasks/Makefile:77]: it is not currently building without mkdir -p ?
Secondary expansion is used to be able to use make auto variables in the requirements

File: FreeRTOS/Test/CMock/Makefile [hunk 2]
```diff
@@ -38,18 +42,29 @@ clean:
...
+	@echo -e 'run               : will build and run all $(UNITS) and their various configurations.'
+	@echo -e '                    it will also generate a test report in the JUnit format in $(BUILD_DIR)/report.xml'
+	@echo -e 'run_col           : same as run but the results are in colors(pass: green, fail: red)'
+	@echo -e 'run_formatted     : will show test case results in a formatted way'
+	@echo -e 'run_col_formatted : same as formatted but will show the results in colors'
+	@echo -e 'coverage          : will run code coverage and generate html docs in $(BUILD_DIR)/coverage/index.html'
+	@echo -e 'all               : will build documentations and coverage, which builds and runs all tests'
+
+$(LIB_DIR)/libcmock.so : $(CMOCK_SRC_DIR)/cmock.c                              \
+                         $(CMOCK_SRC_DIR)/cmock.h                              \
```
[the remark this answers is not in the document]
Comment (alfred2g) [reply FreeRTOS/Test/CMock/Makefile:59]: we are using tabs in the recipe, these spaces are in the requirements

[the remark this answers is not in the document]
Comment (paulbartell) [reply FreeRTOS/Test/CMock/Makefile:59]: Understood. It's quite confusing having both types of white space used for indentation in the same file.

File: FreeRTOS/Test/CMock/tasks/Makefile (hunk 1 above)
Comment (paulbartell) [reply FreeRTOS/Test/CMock/tasks/Makefile:77]: Yes. I had quite a bit of trouble getting this branch to build locally without hacking up the Makefiles.

File: FreeRTOS/Test/CMock/tasks/Makefile [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+.SECONDARY:
+# Generate test_runner.c which contains the main function from the test file
+$(SCRATCH_DIR)/%/test_runner.c : $(PROJECT_DIR)/$(PROJECT)_utest_%.c           \
+                                 Makefile | directories
+	ruby $(UNITY_BIN_DIR)/generate_test_runner.rb $(EXEC)_$*.c             \
+                       $(PROJECT_DIR)/$(PROJECT).yml $@
+
+# Create needed directories
+directories :
+	-mkdir   $(SCRATCH_DIR)
```
[the remark this answers is not in the document]
Comment (alfred2g) [reply FreeRTOS/Test/CMock/tasks/Makefile:164]: yeah kept it like that to be able to see exactly what i am creating, we could add the -p once this is stable

File: FreeRTOS/Test/CMock/tasks/Makefile (hunk 1 above)
Comment (alfred2g) [reply FreeRTOS/Test/CMock/tasks/Makefile:77]: can you incorporate the changes to the makefile ? or is it just the -p in makedir ?

**Review by n9wxu — APPROVED** [review]

**Pull request closed without merging**